lib/IO/Socket/Async/SSL.pm6

```diff
@@ -312,22 +317,28 @@ class IO::Socket::Async::SSL {
     method upgrade-client(IO::Socket::Async::SSL:U: IO::Socket::Async:D $conn,
                           :$enc = 'utf8', OpenSSL::ProtocolVersion :$version = -1,
                           :$ca-file, :$ca-path, :$insecure, :$alpn,
-                          Str :$ciphers, Str :$host --> Promise) {
+                          Str :$ciphers, Str :$host, :$sslkeylogfile --> Promise) {
         self!client-setup:
             { Promise.kept($conn) },
             :$enc, :$version, :$ca-file, :$ca-path, :$insecure,
-            :$alpn, :$ciphers, :$host;
+            :$alpn, :$ciphers, :$host, :$sslkeylogfile;
      }
 
      method !client-setup(&connection-source,
                           :$enc = 'utf8', OpenSSL::ProtocolVersion :$version,
                           :$ca-file, :$ca-path, :$insecure, :$alpn, :$ciphers,
-                          Str :$host) {
+                          Str :$host, :$sslkeylogfile) {
         start {
             my $sock = await connection-source();
             my $connected-promise = Promise.new;
             $lib-lock.protect: {
                 my $ctx = self!build-client-ctx($version);
+                if defined($sslkeylogfile) {
+                    SSL_CTX_set_keylog_callback($ctx, sub ($ssl, $line) {
+                        $sslkeylogfile.IO.spurt: $line ~ "\n", :append;
```

Comment:
NativeCall callbacks should in general catch any exceptions that may be thrown, since there's no way to have them thrown over the C stack.

Reply:
Done.